src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -2748,137 +2746,62 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
-      val samDefParams = map2(fun.vparams, samMethType.paramTypes) { (vd, tp) => ValDef(vd.mods, vd.name, TypeTree(tp), vd.rhs) }
-
-      // `final override def ${sam.name}($p1: $T1', ..., $pN: $TN'): ${samMethTp.finalResultType} = ${sam.name}\$body'($p1, ..., $pN)`
-      val samDef =
-        DefDef(Modifiers(FINAL | OVERRIDE | SYNTHETIC),
-          sam.name.toTermName,
-          Nil,
-          List(samDefParams),
-          TypeTree(samMethType.finalResultType) setPos sampos,
-          Apply(Ident(bodyName), fun.vparams.map(gen.paramToArg))
-        )
+        // solve constraints tracked by tvars
+        val targs = solvedTypes(tvars, tparams, variances, upper = false, lubDepth(sam.info :: Nil))
 
-      val samSubclassName = tpnme.ANON_FUN_NAME
-      val classDef =
-        ClassDef(Modifiers(FINAL), samSubclassName, tparams = Nil,
-          gen.mkTemplate(
-            parents    = TypeTree(ptFullyDefined) :: (if (typeIsSubTypeOfSerializable(pt)) Nil else List(TypeTree(SerializableTpe))),
-            self       = noSelfType,
-            constrMods = NoMods,
-            vparamss   = ListOfNil,
-            body       = List(samBodyDef, samDef),
-            superPos   = sampos
-          )
-        )
+        debuglog(s"sam infer: $pt --> ${appliedType(samTyCon, targs)} by ${fun.tpe} <:< $samInfoWithTVars --> $targs for $tparams")
 
-      // type checking the whole block, so that everything is packaged together nicely
-      // and we don't have to create any symbols by hand
-      val block =
-        typedPos(sampos, mode, pt) {
-          Block(classDef, Apply(Select(New(Ident(samSubclassName)), nme.CONSTRUCTOR), Nil))
+        // a fully defined samClassTp
+        val ptFullyDefined = appliedType(samTyCon, targs)
+        if (funTpMatchesExpected(ptFullyDefined)) {
+          debuglog(s"sam fully defined expected type: $ptFullyDefined from $pt for ${fun.tpe}")
+          fun.setType(ptFullyDefined)
+        } else {
+          debuglog(s"Could not define type $pt using ${fun.tpe} <:< ${pt memberInfo sam} (for $sam)")
+          EmptyTree
         }
-
-      // fix owner for parts of the function we reused now that samBodyDef has a symbol
-      samBodyDef.rhs.changeOwner((fun.symbol, samBodyDef.symbol))
-      samBodyDef.rhs.substituteSymbols(fun.vparams.map(_.symbol), samBodyDef.vparamss.head.map(_.symbol))
-
-      // TODO: improve error reporting -- when we're in silent mode (from `silent(_.doTypedApply(tree, fun, args, mode, pt)) orElse onError`)
-      // the errors in the function don't get out...
-      if (block exists (_.isErroneous))
-        context.error(fun.pos, s"Could not derive subclass of $pt\n (with SAM `${sam.defStringSeenAs(ptFullyDefined)}`)\n based on: $fun.")
-
-      classDef.symbol addAnnotation SerialVersionUIDAnnotation
-      block
+      } catch {
+        case e@(_: NoInstance | _: TypeError) => // TODO: we get here whenever pt contains a wildcardtype???
```

Comment:
Will see what I can do. The TODO can go -- a left over from when I was using =:= to compare pt and ptVars, which balks at embedded WildcardTypes.